Source/WebCore/dom/Element.cpp

```diff
@@ -5066,8 +5066,39 @@ void Element::requestFullscreen(FullscreenOptions&& options, RefPtr<DeferredProm
 #else
     bool optionsEnabled = document().settings().fullScreenKeyboardLock();
 #endif
-    if (optionsEnabled && document().page() && document().page()->hardwareKeyboardAttached())
+
+    RefPtr page = document().page();
+    if (optionsEnabled && page && page->chrome().client().isHardwareKeyboardAttached()) {
+        auto& client = document().page()->chrome().client();
+
+        // Registers a callback to exit fullscreen mode
+        client.addHardwareKeyboardAttachmentObserver([weakThis = WeakPtr { *this }](bool attached) {
```

Comment:
Wrong direction with this one, you don't need to teach `Client`/`WebPage` about the observers, you can add it directly to `WebCore::Page`.

i.e. `document().page()->addHardwareKeyboardAttachmentObserver(...)`